components/codec/src/columnar/bytes.rs

```diff
@@ -52,19 +134,90 @@ impl<'a> ValuesEncoder<&'a [u8]> for ValuesEncoderImpl {
     }
 }
 
+fn encode_with_compress<'a, B, I>(buf: &mut B, values: I) -> std::io::Result<()>
+where
+    B: BufMut,
+    I: Iterator<Item = &'a [u8]>,
+{
+    let writer = WriterOnBufMut { buf };
+    let mut enc = Lz4Encoder::new(writer);
+    for v in values {
+        enc.write_all(v)?;
+    }
+    enc.finish()?;
+
+    Ok(())
+}
+
+fn decode_without_compression<B, F>(
+    length_block_buf: &mut B,
+    data_block_buf: &[u8],
+    mut f: F,
+) -> Result<()>
+where
+    B: Buf,
+    F: FnMut(Bytes) -> Result<()>,
+{
+    let mut offset = 0;
+    while length_block_buf.remaining() > 0 {
+        let length = varint::decode_uvarint(length_block_buf).context(Varint)? as usize;
+        let b = Bytes::copy_from_slice(&data_block_buf[offset..offset + length]);
+        f(b)?;
+        offset += length;
+    }
+
+    Ok(())
+}
+
+fn decode_with_compression<F>(
+    mut length_block_buf: &[u8],
+    compressed_data_block_buf: &[u8],
+    reused_buf: &mut Vec<u8>,
+    f: F,
+) -> Result<()>
+where
+    F: FnMut(Bytes) -> Result<()>,
+{
+    let mut decoder = Lz4Decoder::new(compressed_data_block_buf);
+    decoder.read_to_end(reused_buf).context(Decompress)?;
+    decode_without_compression(&mut length_block_buf, &reused_buf[..], f)
+}
```

Comment:
Why not defining as `Encoding`'s internal method? And just expose the `encode` method?